src/vs/workbench/services/files/node/fileService.ts

```diff
@@ -348,24 +371,18 @@ export class FileService implements IFileService {
 
 				// 3.) check to add UTF BOM
 				return addBomPromise.then(addBom => {
-					let writeFilePromise: TPromise<void>;
-
-					// Write fast if we do UTF 8 without BOM
-					if (!addBom && encodingToWrite === encoding.UTF8) {
-						writeFilePromise = pfs.writeFile(absolutePath, value, encoding.UTF8);
-					}
-
-					// Otherwise use encoding lib
-					else {
-						const encoded = encoding.encode(value, encodingToWrite, { addBOM: addBom });
-						writeFilePromise = pfs.writeFile(absolutePath, encoded);
-					}
-
-					// 4.) set contents
-					return writeFilePromise.then(() => {
+					// 4.) set contents and resolve
+					return this.setContentsAndResolve(resource, absolutePath, value, addBom, encodingToWrite, { mode: 0o666, flag: 'w' }).then(undefined, error => {
+						// Can't use 'w' for hidden files, so truncate and use 'r+' if the file exists
+						if (!exists || error.code !== 'EPERM' || !isWindows) {
+							throw error;
+						}
 
-						// 5.) resolve
-						return this.resolve(resource);
+						// 5.) truncate
```

Comment:
I suggest to add a little bit more comment here why the trick of `truncate` and `r+` mode actually works for hidden files. Currently it is hard to understand what is going on in the error case.